parens.go

```diff
@@ -8,32 +8,30 @@ import (
 
 // New returns a new root context initialised based on given options.
 func New(opts ...Option) *Context {
```

Comment:
Agreed. I am just bit skeptical of value types sometime since all the edge cases it can cause (https://github.com/golang/go/wiki/CodeReviewComments#receiver-type).. 

But i guess here it would be better to use Value type.

Reply:
What do you suggest we do with `Analyzer.Analyze(ctx *Context....)`  and other similar signatures ?

Reply:
I think they should take concrete `Context` values.  In fact, I think everything should take concrete values for `Context` _except_ `parens.Option`, which obviously needs a pointer.

Reply:
this needs some more thought i think 🤔 . 

1. `Expr` type anyways needs to take a pointer to `Env` since it might need to do push a stack frame etc.  
2. Analyser and Expander, I am not sure if this can be a scenario. let me know if you think this is valid: some Exprs might need to maintain a pointer to the env they were analysed/expanded in so that when they are evaluated, they can generate a side effect in those envs..

Reply:
>Expr type anyways needs to take a pointer to Env since it might need to do push a stack frame etc.

Ah, right of course.  In that case, we should probably just pass a `*Context` everywhere.